File: 23. Merge k Sorted Lists/23. Merge k Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

/**
 * Runtime: 4 ms, faster than 78.99% of Java online submissions for Merge k Sorted Lists.
 * Memory Usage: 44.6 MB, less than 21.32% of Java online submissions for Merge k Sorted Lists.
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode dummy = new ListNode(0);
        ListNode output = dummy;
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a,b) -> a.val - b.val);

        for (ListNode l: lists) {
            if (l != null) {
                minHeap.add(l);
            }
        }

        while (!minHeap.isEmpty()) {
            ListNode n = minHeap.poll();
            if(n.next != null) {
                minHeap.add(n.next);
            }
            output.next = n;
            output = output.next;
        }
        return dummy.next;
    }
}

File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
            
        ListNode.__gt__ = lambda self, other : self.val > other.val
    
        dummy = current = ListNode()
        heap = []
        
        for l in lists:
            if l:
                heapq.heappush(heap, (l.val, l))
        
        while len(heap):
            popped = heapq.heappop(heap)
            current.next = popped[1]
            current = current.next
            
            if popped[1].next:
                heapq.heappush(heap, (popped[1].next.val, popped[1].next))
        
        return dummy.next